File: coding_int/chapter_2/2_4_Partition.py
class node:
    def __init__(self, val = None):
        self.next = None
        self.val = val

class llist:
    def __init__(self, head_node = None):
        self.head = head_node
        self.tail = None

    def __iter__(self):
        current = self.head
        while current is not None:
            yield current
            current = current.next

    def __str__(self):
        print("---start---")
        for item in self:
            if item.val is not None:
                print("{0} \t".format(item.val))

        return "---end---"
    
    def addValue(self, value):

        new_node = node(value)
        if self.head is None:
            self.head = new_node
            self.tail = new_node
        else:
            self.tail.next = new_node
            self.tail = new_node
    
    def partition(self, pivot):

        less = llist()
        more = llist()

        for item in self:
            if item.val is not None:
                if (item.val < pivot):
                    less.addValue(item.val)
                elif item.val >= pivot:
                    more.addValue(item.val)

        self.head = less.head
        self.tail = less.tail
        self.tail.next = more.head
        self.tail = more.tail

if __name__ == "__main__":

    test_list = llist()
    test_list.addValue(1)
    test_list.addValue(5)
    test_list.addValue(6)
    test_list.addValue(9)
    test_list.addValue(3)
    test_list.addValue(7)

    print(test_list)

    test_list.partition(6)
    print(test_list)
    

    

File: coding_int/chapter_2/2_4_Partition.cpp
#include<iostream>

class node {
    public:

    node* next;
    int data;
};

class llist {

    private:

    node* head;
    node* tail;

    public:

    llist() {
        head = NULL;
        tail = NULL;
    }

    void addVal(int newdata) {

        node *temp = new node;
        temp->data =  newdata;
        temp->next = NULL;
        
        if (head==NULL) {
            head = temp;
            tail = temp;
        } else {
            tail->next = temp;
            tail = temp;
        }
    }

    void printList() {

        node *temp = new node;
        temp = head;

        while(temp) {
            std::cout << temp->data << "\t";
            temp = temp->next;
        }
        std::cout << std::endl << "---end---" << std::endl; 

    }

    void partition(const int pivot) {

        // left - less than pivot
        // right - equal to or greater than pivot

        llist *less = new llist;
        llist *more = new llist;

        node *temp = new node;
        temp = head;

        while(temp){
            
            if (temp->data < pivot) {
                less->addVal(temp->data);
            } else if (temp->data >= pivot) {
                more->addVal(temp->data);
            }

            temp = temp->next;
        }

        head = less->head;
        tail = less->tail;
        tail->next = more->head;
        tail = more->tail;
    }
};

int main() {

    llist *testing = new llist;
    testing->addVal(1);
    testing->addVal(5);
    testing->addVal(6);
    testing->addVal(9);
    testing->addVal(3);
    testing->addVal(7);

    testing->printList();

    testing->partition(6);
    testing->printList();

    testing->partition(3);
    testing->printList();

}